File: app/src/test/java/acr/browser/lightning/utils/SubscriptionUtilsTest.java
package acr.browser.lightning.utils;

import com.anthonycr.bonsai.Subscription;

import org.junit.Assert;
import org.junit.Test;

/**
 * Unit tests for {@link SubscriptionUtils}.
 */
public class SubscriptionUtilsTest {

    @Test
    public void safeUnsubscribe_NullSubscription_Succeeds() {
        SubscriptionUtils.safeUnsubscribe(null);
    }

    @Test
    public void safeUnsubscribe_NonNullSubscription_SuccessfullyUnsubscribes() {
        Subscription subscription = new Subscription() {

            boolean isUnsubscribed = false;

            @Override
            public void unsubscribe() {
                isUnsubscribed = true;
            }

            @Override
            public boolean isUnsubscribed() {
                return isUnsubscribed;
            }
        };

        Assert.assertFalse(subscription.isUnsubscribed());

        SubscriptionUtils.safeUnsubscribe(subscription);

        Assert.assertTrue(subscription.isUnsubscribed());
    }
}

File: app/src/test/java/acr/browser/lightning/utils/SubscriptionUtilsTest.kt
package acr.browser.lightning.utils

import com.anthonycr.bonsai.Subscription

import org.junit.Assert
import org.junit.Test

/**
 * Unit tests for [SubscriptionUtils].
 */
class SubscriptionUtilsTest {

    @Test
    fun `safeUnsubscribe succeeds with null subscription`() = SubscriptionUtils.safeUnsubscribe(null)

    @Test
    fun `safeUnsubscribe unsubscribes successfully with valid description`() {
        val subscription = object : Subscription {

            internal var isUnsubscribed = false

            override fun unsubscribe() {
                isUnsubscribed = true
            }

            override fun isUnsubscribed(): Boolean = isUnsubscribed
        }

        Assert.assertFalse(subscription.isUnsubscribed())

        SubscriptionUtils.safeUnsubscribe(subscription)

        Assert.assertTrue(subscription.isUnsubscribed())
    }
}